File: clients/csharp/Codegame/ServerMessage.cs
namespace SpbAiChamp.Codegame
{
    /// <summary>
    /// Message sent from server
    /// </summary>
    public abstract class ServerMessage
    {
        /// <summary> Write ServerMessage to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read ServerMessage from reader </summary>
        public static ServerMessage ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case GetAction.TAG:
                    return GetAction.ReadFrom(reader);
                case Finish.TAG:
                    return Finish.ReadFrom(reader);
                case DebugUpdate.TAG:
                    return DebugUpdate.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// Get action for next tick
        /// </summary>
        public class GetAction : ServerMessage
        {
            public const int TAG = 0;
        
            /// <summary>
            /// Player's view
            /// </summary>
            public Model.Game PlayerView { get; set; }
            /// <summary>
            /// Whether app is running with debug interface available
            /// </summary>
            public bool DebugAvailable { get; set; }
        
            public GetAction() { }
        
            public GetAction(Model.Game playerView, bool debugAvailable)
            {
                this.PlayerView = playerView;
                this.DebugAvailable = debugAvailable;
            }
        
            /// <summary> Read GetAction from reader </summary>
            public static new GetAction ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new GetAction();
                result.PlayerView = Model.Game.ReadFrom(reader);
                result.DebugAvailable = reader.ReadBoolean();
                return result;
            }
        
            /// <summary> Write GetAction to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                PlayerView.WriteTo(writer);
                writer.Write(DebugAvailable);
            }
        
            /// <summary> Get string representation of GetAction </summary>
            public override string ToString() {
                string stringResult = "GetAction { ";
                stringResult += "PlayerView: ";
                stringResult += PlayerView.ToString();
                stringResult += ", ";
                stringResult += "DebugAvailable: ";
                stringResult += DebugAvailable.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Signifies end of the game
        /// </summary>
        public class Finish : ServerMessage
        {
            public const int TAG = 1;
        
        
            public Finish() { }
        
            /// <summary> Read Finish from reader </summary>
            public static new Finish ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Finish();
                return result;
            }
        
            /// <summary> Write Finish to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
            }
        
            /// <summary> Get string representation of Finish </summary>
            public override string ToString() {
                string stringResult = "Finish { ";
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Debug update
        /// </summary>
        public class DebugUpdate : ServerMessage
        {
            public const int TAG = 2;
        
            /// <summary>
            /// Player's view
            /// </summary>
            public Model.Game PlayerView { get; set; }
        
            public DebugUpdate() { }
        
            public DebugUpdate(Model.Game playerView)
            {
                this.PlayerView = playerView;
            }
        
            /// <summary> Read DebugUpdate from reader </summary>
            public static new DebugUpdate ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new DebugUpdate();
                result.PlayerView = Model.Game.ReadFrom(reader);
                return result;
            }
        
            /// <summary> Write DebugUpdate to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                PlayerView.WriteTo(writer);
            }
        
            /// <summary> Get string representation of DebugUpdate </summary>
            public override string ToString() {
                string stringResult = "DebugUpdate { ";
                stringResult += "PlayerView: ";
                stringResult += PlayerView.ToString();
                stringResult += " }";
                return stringResult;
            }
        }
    }
}

File: clients/fsharp/Codegame/ServerMessage.fs
#nowarn "0058"

namespace SpbAiChamp
namespace SpbAiChamp.Codegame

open SpbAiChamp

/// Get action for next tick
type ServerMessageGetAction = {
    /// Player's view
    PlayerView: Model.Game;
    /// Whether app is running with debug interface available
    DebugAvailable: bool;
} with

    /// Write GetAction to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        this.PlayerView.writeTo writer
        writer.Write this.DebugAvailable
        ()

    /// Read GetAction from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        PlayerView = Model.Game.readFrom reader;
        DebugAvailable = reader.ReadBoolean()
    }

/// Signifies end of the game
type ServerMessageFinish = struct end with

    /// Write Finish to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        ()

    /// Read Finish from reader
    static member readFrom(reader: System.IO.BinaryReader) = new ServerMessageFinish()

/// Debug update
type ServerMessageDebugUpdate = {
    /// Player's view
    PlayerView: Model.Game;
} with

    /// Write DebugUpdate to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 2
        this.PlayerView.writeTo writer
        ()

    /// Read DebugUpdate from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        PlayerView = Model.Game.readFrom reader;
    }

/// Message sent from server
type ServerMessage =
    /// Get action for next tick
    | GetAction of ServerMessageGetAction
    /// Signifies end of the game
    | Finish of ServerMessageFinish
    /// Debug update
    | DebugUpdate of ServerMessageDebugUpdate
    with

    /// Write ServerMessage to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | GetAction value -> value.writeTo writer
            | Finish value -> value.writeTo writer
            | DebugUpdate value -> value.writeTo writer

    /// Read ServerMessage from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> GetAction (ServerMessageGetAction.readFrom reader)
            | 1 -> Finish (ServerMessageFinish.readFrom reader)
            | 2 -> DebugUpdate (ServerMessageDebugUpdate.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)